src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DelayUnassignedAllocation.java

```diff
@@ -0,0 +1,307 @@
+                    for (Iterator<MutableShardRouting> it = unassigned.iterator(); it.hasNext(); ) {
+                        MutableShardRouting unassignedShard = it.next();
+                        if (unassignedShard.shardId().equals(shard.shardId())) {
+                            it.remove();
+                            allocation.routingNodes().ignoredUnassigned().add(unassignedShard);
+                            break;
+                        }
+                    }
+                }
+            }
+
+            allocation.routingNodes().unassigned().addAll(unassigned);
+        }
+
+        return new Result(changed, scheduleReroute ? duration : null);
+    }
+
+    /**
+     * Result of delayed allocation.
+     */
+    public static class Result {
+        private final boolean changed;
+        private final TimeValue rerouteSchedule;
+
+        public Result(boolean changed, TimeValue rerouteSchedule) {
+            this.changed = changed;
+            this.rerouteSchedule = rerouteSchedule;
+        }
+
+        /**
+         * Has anything changed in the allocation.
+         */
+        public boolean isChanged() {
+            return this.changed;
+        }
+
+        /**
+         * Is reroute needed in order to act upon node delayed duration expiry.
+         */
+        public boolean isRerouteRequired() {
+            return rerouteSchedule != null;
+        }
+
+        /**
+         * The reroute schedule to execute to act upon node delayed duration expiry.
+         */
+        public TimeValue getRerouteSchedule() {
+            return this.rerouteSchedule;
+        }
+    }
+
+    static class DelayedAllocation {
+
+        public final DiscoveryNode node;
+        public final List<ShardRouting> shards;
+        public final long failedTimestampInNanos;
+
+        public DelayedAllocation(DiscoveryNode node, List<ShardRouting> shards, long failedTimestampInNanos) {
+            this.node = node;
+            this.shards = shards;
+            this.failedTimestampInNanos = failedTimestampInNanos;
+        }
+    }
+
+    public enum NodeKey {
+        NAME() {
+            @Override
+            public Object getNodeKey(DiscoveryNode node) {
+                return node.getName();
+            }
+        },
+        ID() {
+            @Override
+            public Object getNodeKey(DiscoveryNode node) {
+                return node.getId();
+            }
+        },
+        HOST_ADDRESS() {
+            @Override
+            public Object getNodeKey(DiscoveryNode node) {
+                return node.getHostAddress();
+            }
+        },
+        HOST_NAME() {
+            @Override
+            public Object getNodeKey(DiscoveryNode node) {
+                return node.getHostName();
+            }
+        },
+        TRANSPORT_ADDRESS() {
+            @Override
+            public Object getNodeKey(DiscoveryNode node) {
+                return node.getAddress();
```

Comment:
It may be worthwhile to add `equals` and `hashCode` to the `TransportAddress` interface so that any future implementation does not skip it, excepting the singleton variants like the Dummy one (mostly worried that `BoundTransportAddress` may one day implement the interface but not implement them).